js/walutomat.js

```diff
@@ -0,0 +1,267 @@
+            'api': {
+                'public': {
+                    'get': [
+                        'public/market/orderbook/{symbol}',
+                    ],
+                },
+                'private': {
+                    'get': [
+                        'account/id',
+                        'account/balances',
+                        'account/history',
+                        'market/orders',
+                        'market/orders/{orderId}',
+                    ],
+                    'post': [
+                        'market/orders',
+                        'market/orders/close/{orderId}',
+                    ],
+                },
+            },
+            'exceptions': {
+                'INVALID ARGUMENT': ExchangeError,
+                'INVALID PARAMETER': ExchangeError,
+                'INSUFFICIENT_FUNDS': InsufficientFunds,
+                'MISSING_API_KEY': AuthenticationError,
+            },
+        });
+    }
+
+    fetchMarkets (params = {}) {
+        let currencies = ['EUR', 'GBP', 'USD', 'CHF', 'PLN'];
+        let markets = [];
+        let currenciesLength = currencies.length;
+        for (let base = 0; base < currenciesLength - 1; base++) {
+            let newIndex = base + 1;
+            for (let quote = newIndex; quote < currencies.length; quote++) {
+                markets.push ({
+                    'id': currencies[base] + '_' + currencies[quote],
+                    'symbol': currencies[base] + '/' + currencies[quote],
+                    'base': currencies[base],
+                    'quote': currencies[quote],
+                    'active': true,
+                    'precision': {
+                        'amount': 2,
+                        'price': 4,
+                    },
+                    'info': currencies[base] + ' - ' + currencies[quote] + ' exchange',
+                });
+            }
+        }
+        return markets;
+    }
+
+    async fetchBalance (params = {}) {
+        let balances = await this.privateGetAccountBalances ();
+        let result = {};
+        for (let i = 0; i < balances.length; i++) {
+            let balance = balances[i];
+            result[this.safeString (balance, 'currency')] = {
+                'free': this.safeFloat (balance, 'balanceAvailable'),
+                'used': this.safeFloat (balance, 'balanceReserved'),
+                'total': this.safeFloat (balance, 'balanceAll'),
+            };
+        }
+        return this.parseBalance (result);
+    }
+
+    async fetchOrderBook (symbol, limit = undefined, params = {}) {
+        await this.loadMarkets ();
+        let symbolObject = { 'symbol': this.marketId (symbol) };
+        let response = await this.publicGetPublicMarketOrderbookSymbol (this.extend (symbolObject, params));
+        return this.parseOrderBook (response, undefined, 'bids', 'asks', 'price', 'baseVolume');
+    }
+
+    async fetchTrades (symbol, since = undefined, limit = undefined, params = {}) {
```

Comment:
Hi, @piotrmoszkowicz! Pardon for the long wait, was occupied with my other work.

> Can we merge it :) ?

Almost!) it still has a few errors that we need to wipe out clean, namely, this particular method, `fetchTrades` is actually calling a private method, so it kinda implements the functionality of `fetchMyTrades` (private) instead of `fetchTrades` (public), and the body of this method doesn't seem to be correct, this particular place:

```
let result = [];
        for (let i = 0; i < response.length; i++) {
            result.push ({
                'id': this.safeInteger (response, 'id'), // ←---- it's missing response[i]
            });
        }
        return result;
```

↑ that line is probably missing an index `i` and even if you put it there – it's very unclear what is the intent/purpose of this loop. Can you rename `fetchTrades` to `fetchMyTrades` (effectively setting `has['fetchTrades'] = false`) and take one more look into the body of this method? There's also a couple of inconsistencies in how it treats the `since` argument and the `limit` argument – first, of it does not filter the results and it puts undefined values into the request, which contradicts to pagination principles as described here: https://github.com/ccxt/ccxt/wiki/Manual#pagination

So, we will definitely have to rework this method, unfortunately, because it wouldn't fit and work as is.

Reply:
Yeah, but in this case we're talking of the difference between `fetchTrades` and `fetchMyTrades`, and they don't have `fetchTrades`, so it has to be renamed and we still to fix the body of the method.

Reply:
Sure, will do, just wanted to confirm 🙂

Reply:
@piotrmoszkowicz thx so much for your effort! Really appreciate your help with this!

On the `symbol`, `since` and `limit`, here's a general pattern that we are following everywhere:
- if the exchange does not require those – we don't set it explicitly in the request, instead we don't send it thus, defaulting to whatever is the exchange's default value
- if `since` is in milliseconds – we implement the date-based pagination, otherwise we don't set it

Examples:

- https://github.com/ccxt/ccxt/blob/master/js/hitbtc2.js#L1188
- https://github.com/ccxt/ccxt/blob/master/js/hitbtc2.js#L1221

Reply:
The problem with this method is that it is not documented in their API. I've gotten information, that it is "like it was before", so the implementation of @jpospychala hasn't been changed by me. I've pushed the changes :)